README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# conservedPos

<!-- badges: start -->

<!-- badges: end -->

## Description

The goal of conservedPos package is to find important conserved sites
for small target region by conducting a site-specific analysis for a set
of aligned DNA sequences. The package will provide function to count
nucleotides occurrances in each position independently and calculates
the frequency of the nucleotides in order to quantify the conservity of
each position.

## Installation

You can install the released version of conservedPos

``` r
require("devtools")
devtools::install_github("hezijin/conservedPos", build_vignettes = TRUE)
library("conservedPos")
```

To run the Shiny app:

``` r
runconservedPos()
```

For tutorial how to use:

``` r
browseVignettes("conservedPos")
```

## Overview

``` r
ls("package:conservedPos")
data(package = "conservedPos")
```

The package contains 5 function to reformat and analyze the input
sequences and 2 function to plot for the result. createPosVec() grabs
the nucleotide for all of sequences at index position. Then in
conservityTable(), it loops all of the index and calculate the most
conserved nucleotide frequency through findConservityFromS(). We can
also use findConservedNul() to find the conserved nucleotide at specific
site and findMaxLen() to find the max length for sequence set.

plotOverall() plots the overall alignment’s conservity using the
conservity vector generated from the function conservityTable while
plotPartial()plots the seleted region from users.

Overview picture for the package:

![](./inst/extdata/conservedPos_overview.jpg)

## Contribution

The idea of developing this package is inspired by PWM Viewing and
Searching. The way of clear up and reformat sequence string in function
createPosVec is referenced to many website (show in Reference section).
The idea of function findConservityFromS comes from the function in
Biostirngs package function NucleotideFrequencyAt, while my function
avoids the limitation of the input (NucleotideFrequencyAt is not allowed
to use stirngs with NA, which means length of the input sequences need
to be equal).The function findConservedNul() uses the idea from stack
overflow.(see reference 8)

The function findMaxLen and conservityTable are written by me. The way
of generate bar plot comes from R language Tutorial(show in Reference
section) and used in function plotOverall and plotPartial.

## Reference

1.  Silva, A. (2020) TestingPackage: An Example R Package For BCB410H.
    Unpublished. URL <https://github.com/anjalisilva/TestingPackage>

2.  R Bar Plot (2018) DataMentor. Getting Started in Data Science With
    R. <https://www.datamentor.io/r-programming/bar-plot/>

3.  Lab 1: Biostrings in R. (2019) Susan Holmes and Wolfgang.
    Huberhttps://web.stanford.edu/class/bios221/labs/biostrings/lab\_1\_biostrings.html

4.  R - Bar Charts. Bar Chart Labels, Title and Colors.(2020)
    Tutorialpoints-R Tutorial.
    <https://www.tutorialspoint.com/r/r_bar_charts.htm>

5.  H. Pagès, P. Aboyoun, R. Gentleman and S. DebRoy (2020). Biostrings:
    Efficient manipulation of biological strings. R package version
    2.58.0. <https://bioconductor.org/packages/Biostrings>

6.  Orchestrating high-throughput genomic analysis with Bioconductor. W.
    Huber, V.J. Carey, R. Gentleman, …, M. Morgan Nature Methods,
    2015:12, 115.

7.  Bembom O, Ivanek R (2020). seqLogo: Sequence logos for DNA sequence
    alignments. R package version 1.56.0.
    <http://www.bioconductor.org/packages/release/bioc/html/seqLogo.html>

8.  Antonis. (2018) How to find the most repeated word in a vector with
    R. Stack oveflow.
    <https://stackoverflow.com/questions/48632957/how-to-find-the-most-repeated-word-in-a-vector-with-r>

## Acknowledgements

This package was developed as part of an assessment for 2020BCB410H:
Applied Bioinformatics, University of Toronto, Toronto,CANADA.


NEWS.md
# conservedPos 0.1.2

* Added more test for functions
* Added more examples
* Change the the way of reading file: readDNAStringSet to readBStringSet, larger range of input choices (DNA/RNA/Protein)


# conservedPos 0.1.1

* Added citation to the package.
* Added shiny app scripts and launcher.

# conservedPos 0.1.0

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(conservedPos)
library(Biostrings)

test_check("conservedPos")


tests/testthat/testPosCal.R
context("Test functions in PositionCal.R")


test_that("create correct Position Vec", {
  data(testSeqShort)
  testSeq <- testSeqShort
  pos <- createPosVec(testSeq, 1)
  expect_equal(pos, "ATT")
})

test_that("find correct conservity from a index 1", {
  data(testSeqShort)
  testSeq <- testSeqShort
  pos <- createPosVec(testSeq, 1)
  cons <- findConservityFromS(pos)
  expect_equal(cons, 2/3)
})

test_that("find correct conservity from a index 12", {
  data(testSeqShort)
  testSeq <- testSeqShort
  pos <- createPosVec(testSeq, 12)
  cons <- findConservityFromS(pos)
  expect_equal(cons, 2/3)
})

test_that("find conserved nucleotide from a index 12", {
  data(testSeqShort)
  testSeq <- testSeqShort
  pos <- createPosVec(testSeq, 12)
  cons <- findConservedNul(pos)
  expect_equal(cons, "A")
})

test_that("find conserved nucleotide from a index 2", {
  data(testSeqShort)
  testSeq <- testSeqShort
  pos <- createPosVec(testSeq, 2)
  cons <- findConservedNul(pos)
  expect_equal(cons, "G")
})

test_that("find conserved nucleotide from a index 20", {
  findpath <- system.file("extdata", "sampleSeq2.txt", package = "conservedPos")
  rawSeq <- Biostrings::readBStringSet(findpath,"fasta")
  pos <- createPosVec(rawSeq, 20)
  cons <- findConservedNul(pos)
  expect_equal(cons, "T")
})

test_that("find correct max length in BStringSet", {
  data(testSeqUnequal)
  testSeq <- testSeqUnequal
  maxL <- findMaxLen(testSeq)
  expect_equal(maxL, 37)
})

test_that("find correct max length in BStringSet", {
  data(testSeqUnequal)
  testSeq <- testSeqUnequal
  maxL <- findMaxLen(testSeq)
  expect_equal(maxL, 37)
})

test_that("find correct max length in BStringSet", {
  data(testSeqShort)
  testSeq <- testSeqShort
  maxL <- findMaxLen(testSeq)
  expect_equal(maxL, 37)
})

test_that("find correct max length in BStringSet", {
  findpath <- system.file("extdata", "sampleSeq1.txt", package = "conservedPos")
  rawSeq <- Biostrings::readBStringSet(findpath,"fasta")
  maxL <- findMaxLen(rawSeq)
  expect_equal(maxL, 28)
})

test_that("find correct max length in BStringSet", {
  findpath <- system.file("extdata", "sampleSeq2.txt", package = "conservedPos")
  rawSeq <- Biostrings::readBStringSet(findpath,"fasta")
  maxL <- findMaxLen(rawSeq)
  expect_equal(maxL, 210)
})

test_that("find correct max length in BStringSet", {
  findpath <- system.file("extdata", "sampleSeq3.txt", package = "conservedPos")
  rawSeq <- Biostrings::readBStringSet(findpath,"fasta")
  maxL <- findMaxLen(rawSeq)
  expect_equal(maxL, 140)
})

test_that("check the conservity vector value at several index", {
  data(testSeqUnequal)
  testSeq <- testSeqUnequal
  conserV <- conservityTable(testSeq)

  expect_equal(conserV[37], 1)
  expect_equal(conserV[1], 2/3)
  expect_equal(conserV[2], 1)
})

test_that("check the conservity vector value at several index", {
  findpath <- system.file("extdata", "sampleSeq1.txt", package = "conservedPos")
  rawSeq <- Biostrings::readBStringSet(findpath,"fasta")
  conserV <- conservityTable(rawSeq)

  expect_equal(conserV[11], 0.5)
  expect_equal(conserV[5], 2/3)
  expect_equal(conserV[10], 0.5)
})



tests/testthat/test-testIndexInput.R
context("Test function plotPartial index validity")

test_that("Input Index out of range", {
  data(testSeqUnequal)
  testSeq <- testSeqUnequal
  conserV <- conservityTable(testSeq)

  message <- plotPartial(conserV, 1, 100)
  expect_equal(message, "Invalid Index!")
  message <- plotPartial(conserV, 100, 1)
  expect_equal(message, "Invalid Index!")
})

test_that("first Index is larger", {
  data(testSeqUnequal)
  testSeq <- testSeqUnequal
  conserV <- conservityTable(testSeq)

  message <- plotPartial(conserV, 12, 5)
  expect_equal(message, "Invalid Index!")
})



R/plotConser.R
#' Plot the conservity of the positions
#'
#'
#' This function plots the overall alignment's conservity
#' using the conservity vector generated from the function
#' conservityTable.
#'
#' @param conserV a vector with serveral number of conservity
#'
#' @return a plot to view the overall alignment's conservity
#'
#' @examples
#'
#' # Example 1: Plot a conservity against of each position for testSeqShort
#'
#' data(testSeqShort)
#' testSeq <- testSeqShort
#'
#' # Use conservityTable() to create table for testSeqShort,
#' # and pass it to plotOverall() for plotting
#' table  <- conservityTable(testSeq)
#' oPar <- par(mar = c(2, 4, 2, 1)) # Set plot parameters to layout plot in center.
#' plotOverall(table)
#'
#' # Example 2: Plot a conservity against of each position for raw data
#' # sampleSeq3.txt. This should takes longer time to run.
#' findpath <- system.file("extdata", "sampleSeq3.txt", package = "conservedPos")
#' rawSeq <- Biostrings::readBStringSet(findpath,"fasta")
#'
#' # Use conservityTable() to create table for rawSeq,
#' # and pass it to plotOverall() for plotting
#' table <- conservityTable(rawSeq)
#' oPar <- par(mar = c(2, 4, 2, 1)) # Set plot parameters to layout plot in center.
#' plotOverall(table)
#'
#' @references
#' R - Bar Charts: Bar:Chart Labels, Title and Colors.
#' \emph{Tutorialpoints-R Tutorial} \href{https://www.tutorialspoint.com/r/r_bar_charts.htm}{Link}
#'
#' Silva, A. (2020) TestingPackage: An Example R Package For BCB410H. Unpublished.
#' \href{https://github.com/anjalisilva/TestingPackage}{Link}
#'
#' R Bar Plot (2018) DataMentor
#' \emph{Getting Started in Data Science With R } \href{https://www.datamentor.io/r-programming/bar-plot/}{Link}
#'
#' @export
#' @importFrom graphics barplot
plotOverall <- function(conserV){
  graphics::barplot(conserV, xlab = "Position", ylab = "Conservity", col = "blue",
          names.arg = c(1:length(conserV)), main="Conservity for Input Sequences")
}

#' This function plots the conservity for selected alignment region
#' from user using the conservity vector.
#'
#' @param conserV a vector with serveral number of conservity
#' @param firstIndex a number to point the start index of selected region
#' @param lastIndex a number to point the end index of selected region
#'
#' @return a plot to view the seleted region alignment's conservity
#'
#' @examples
#' # Example 1: Plot a conservity against of each position for testSeqShort
#' # from position 2 to 15
#' data(testSeqShort)
#' testSeq <- testSeqShort
#' # Use conservityTable() to create table for testSeqShort,
#' # and pass it to plotOverall() for plotting
#' table  <- conservityTable(testSeq)
#' oPar <- par(mar = c(2, 4, 2, 1)) # Set plot parameters to layout plot in center.
#' plotPartial(table, 2, 15)
#'
#' # Example 2: Plot a conservity against of each position for raw data
#' # from position 50 to 100
#' # sampleSeq3.txt. This should takes longer time to run.
#' findpath <- system.file("extdata", "sampleSeq3.txt", package = "conservedPos")
#' rawSeq <- Biostrings::readBStringSet(findpath,"fasta")
#'
#' # Use conservityTable() to create table for rawSeq,
#' # and pass it to plotOverall() for plotting
#' table  <- conservityTable(testSeq)
#' oPar <- par(mar = c(2, 4, 2, 1)) # Set plot parameters to layout plot in center.
#' plotPartial(table, 50, 100)
#'
#'
#' @references
#'R - Bar Charts: Bar Chart Labels, Title and Colors.
#' \emph{Tutorialpoints-R Tutorial} \href{https://www.tutorialspoint.com/r/r_bar_charts.htm}{Link}
#'
#' Silva, A. (2020) TestingPackage: An Example R Package For BCB410H. Unpublished.
#' \href{https://github.com/anjalisilva/TestingPackage}{Link}
#'
#' R Bar Plot (2018) DataMentor
#' \emph{Getting Started in Data Science With R } \href{https://www.datamentor.io/r-programming/bar-plot/}{Link}
#'
#' @export
#' @importFrom graphics barplot
plotPartial <- function(conserV, firstIndex, lastIndex){
  maxL <- length(conserV)
  if(firstIndex > maxL | lastIndex > maxL | firstIndex >= lastIndex){
    print("Invalid Index!")
  }else{
    graphics::barplot(conserV[firstIndex:lastIndex], xlab = "Position", ylab = "Conservity", col = "blue",
            names.arg = c(firstIndex:lastIndex), main = "Plot of site conservity")
  }


}



inst/shiny-scripts/app.R
library(shiny)
# This layout of app is adapted from
# app.R section. Tabsets(2020) Shiny in R studio
# https://shiny.rstudio.com/gallery/tabsets.html
# Other Reference:
# Using sliders(2017) Shiny in R studio
# https://shiny.rstudio.com/articles/sliders.html

ui <- fluidPage(
  titlePanel("conservedPos"),

  sidebarLayout(
    sidebarPanel(

      fileInput("file", "File input:"),

      br(),

      sliderInput("range", "Choose plot pange in max length of conservity table",
                  min = 1, max = 200,
                  value = c(1,10)),
      br(),

      textInput("site",
                "Choose a specific site",
                value = 1)


    ),

    mainPanel(
      tabsetPanel(type = "tabs",
                  tabPanel("Conservity Table", verbatimTextOutput("table")),
                  tabPanel("Conservity Plot", plotOutput("plot")),
                  tabPanel("Site-specific Conservity", tableOutput("site"))

      )
    )

  )
)

server <- function(input, output) {


    output$table<- renderPrint({
      testSeq <- Biostrings::readBStringSet(input$file$datapath)
      mytable <- conservityTable(testSeq)
      cat("Conservity for uploaded sequences set, max length",findMaxLen(testSeq))
      cat("\n")
      mytable
    })

    output$plot<- renderPlot({
      testSeq <- Biostrings::readBStringSet(input$file$datapath)
      mytable <- conservityTable(testSeq)
      plotPartial(mytable, input$range[1], input$range[2])

    })

    output$site<- renderPrint({
      testSeq <- Biostrings::readBStringSet(input$file$datapath)
      posVec <- createPosVec(testSeq,as.numeric(input$site))
      cat("Nucleotide/Residue with highest frequency at position:", findConservedNul(posVec))
      cat(" with frequency", findConservityFromS(posVec))
      #

    })


  }
shinyApp(ui = ui, server = server)


R/PositionCal.R
#' Find most conserved nucleotide in each position and its frequency
#'
#' This function grab the nucleotide for all of sequences at index position
#' to create a XString sequence for a collection of this nucleotide. Then
#' the function reformats the Xstring into a string clean it up.
#'
#' @param seqSet A set of aligned sequence read from fasta file
#' @param index A numeric index for generating Position String at that position.
#'
#' @return A String containing nucleotides at index position for sequences.
#'
#' @examples
#' # Example 1: Use a sequence set with short sequence from data,
#' # create a string containing nucleotides at index 1 for this sequence set.
#'
#' data(testSeqShort)
#' testSeq <- testSeqShort
#' position1Vec <- createPosVec(testSeq, 1)
#' position1Vec # "ATT"
#'
#' # Example 2: Use a raw data sequence set, create a string containing
#' # nucleotides at index 10 for these raw data sequences.
#' findpath <- system.file("extdata", "sampleSeq3.txt", package = "conservedPos")
#' rawSeq <- Biostrings::readBStringSet(findpath,"fasta")
#' position10Vec <- createPosVec(rawSeq, 10)
#' position10Vec <- "TTTCACTACCAA"
#' @export
createPosVec <- function(seqSet, index){
  position <- c()

  # Loop over each sequences and grab the nucleotide at index postion.
  for (i in 1: length(seqSet)){
    if(length(seqSet[[i]]) >= index){
      nul <- seqSet[[i]][index]
      nul <- paste(nul, "")
      position[i] <- nul
    }
  }

  # Reformat position from vector to a string and remove NA and whitespace
  position <- paste(position, collapse = "")
  position <- gsub(" ", "", position, fixed = TRUE)
  position <- gsub("NA", "", position, fixed = TRUE)
  return (position)
}

#' This function finds the nucleotide that has highest frequency of
#' in input string.
#'
#' @param str A string containing character nucleotides
#'
#' @return  Return a vector that has one or more nucleotides
#'          with highest frequency
#'
#' @examples
#' # Example 1: find the the letter in string
#' # with highest occurrence.
#'
#' str <- "ATTTGGGGCC"
#' findConservedNul(str) # "G"
#'
#' # Example 2: find the highest frequency nucleotide from a
#' # string of nucleotides generated by createPosVec()
#'
#' data(testSeqShort)
#' testSeq <- testSeqShort
#' position1Vec <- createPosVec(testSeq, 1)
#' findConservedNul(position1Vec) # "T"
#'
#' @references
#' Antonis. (2018) How to find the most repeated word in a vector with R
#' \emph{Stack oveflow}  \href{https://stackoverflow.com/questions/48632957/how-to-find-the-most-repeated-word-in-a-vector-with-r
#' }{Link}
#'
#' @export
findConservedNul <- function(str){
  split <- strsplit(str, "")
  nul <- names(which(table(split) == max(table(split))))
  return(nul)

}

#' This function finds the highest frequency of nucleotide in input string.
#'
#' @param str A string containing character nucleotides
#'
#' @return  Return a numeric value for highest frequency
#'
#' @examples
#' # Example 1: find the highest letter frequency in a string
#'
#' str <- "AADDDCCEEEE"
#' freq <- findConservityFromS(str)
#' freq # 0.363636...
#'
#' # Example 2: find the highest frequency of nucleotide from a string of
#' # nucleotide generated by createPosVec()
#' data(testSeqShort)
#' position1Vec <- createPosVec(testSeqShort, 1)
#' freq <- findConservityFromS(position1Vec)
#' freq # 0.66666...
#' @export
findConservityFromS <- function(str){

  # Split the str into character vector and find max occurances
  # of certain nucleotide.
  split <- strsplit(str, "")
  maxval <- max(table(split))

  # Use the length of the str to calculate the highest frequency
  # of the nulceotide.
  length <- nchar(str)
  return(maxval / length)
}

#' This function finds the max length of sequence for input sequence set
#'
#' @param seqSet seqSet A set of aligned sequence read from fasta file
#'
#' @return A numeric value of the max length of sequence for seqSet
#'
#' @examples
#' # Example 1: find the max length of a sequence set retrieved from raw data
#' findpath <- system.file("extdata", "sampleSeq3.txt", package = "conservedPos")
#' rawSeq <- Biostrings::readBStringSet(findpath,"fasta")
#' maxL <- findMaxLen(rawSeq)
#' maxL # 140
#'
#' # Example 2: find the max length of a sequence set retrieved from data
#' data(testSeqShort)
#' testSeq <- testSeqShort
#' maxL <- findMaxLen(testSeq)
#' maxL # 37
#'
#' @export

findMaxLen <- function(seqSet){
  max <- 0
  for (i in 1:length(seqSet)){
    if (max < length(seqSet[[i]])){
      max <- length(seqSet[[i]])
    }
  }
  return(max)
}

#' This function calculates the conservity of all positions for input
#' sequences and generate a vector contains each position's conservity
#'
#' @param seqSet seqSet A set of aligned sequence read from fasta file
#'
#' @return a vector contains each position's highest frequency
#'
#' @examples
#' # Example 1 : Create conservity Table for testSeqShort.
#' data(testSeqShort)
#' testSeq <- testSeqShort
#' table  <- conservityTable(testSeq)
#' # You will see a vector containing 37 numbers (max length), these numbers are
#' # the highest frequency for each position.
#' table
#'
#' # Example 2: Create conservity Table for another sequence set, a raw data
#' # sampleSeq3.txt. This should takes longer time to run.
#' findpath <- system.file("extdata", "sampleSeq3.txt", package = "conservedPos")
#' rawSeq <- Biostrings::readBStringSet(findpath,"fasta")
#' table <- conservityTable(rawSeq)
#' # Should have 140 numbers(max length of rawSeq), and
#' # each number is the the highest frequency for each position
#' table
#'
#' @export
conservityTable <- function(seqSet){
  # Find max length of sequences in seqSet, it is used as the max range of
  # position to compute the conservity. It also guarantees that each position
  # to be calculated even the sequences have slightly length differences.
  max <- findMaxLen(seqSet)

  # From index 1 to max length, create a position string for each index
  # and find the conservity from that index.
  conserVec <- c()
  for (i in 1:max){
    newstr <- createPosVec(seqSet, i)
    conser <- findConservityFromS(newstr)
    conserVec[i] <- conser
  }
  return(conserVec)
}
